File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.*;

class Solution {
    boolean solution(String s) {
        boolean answer = true;
        
        boolean flag = true;
        
        Stack<Character> stack = new Stack<>();
        
        for (char c : s.toCharArray()) {
            if (stack.isEmpty()) {
                stack.push(c);
            } else {
                if (c == ')') {
                    if (stack.isEmpty()) {
                        flag = false;
                        break;
                    } else {
                        stack.pop();
                    }
                } else {
                    stack.push(c);
                }
            }
        }
        
        if (stack.isEmpty() && flag) {
            answer = true;
        } else {
            answer = false;
        }
        
    

        return answer;
    }
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.py
def solution(s):
    answer = False
    
    stack = []

    for data in s:
        if data == '(':
            stack.append(data)
        else:
            if stack:
                if stack[-1] == '(':
                    stack.pop()
                else:
                    stack.append(data)
            else:
                stack.append(data)

    if not stack:
        answer = True

    return answer